### PR TITLE
External triggering 

### PR DESCRIPTION
**Pull request by adisuciu**
With the new firmware, support for trigger out and cross instrument triggering is added. This pull request makes the necessary changes in Scopy to support that.

### CODE REVIEW

File: src/hardware_trigger.cpp [hunk 1]
```diff
@@ -183,6 +203,49 @@ HardwareTrigger::condition HardwareTrigger::digitalCondition(uint chnIdx)
...
 }
 
+
+
+bool HardwareTrigger::hasExternalTriggerOut()
+{
+	auto ret = iio_channel_find_attr(m_logic_channels[1], "out_select");
+
+	return (ret==NULL) ? false : true;
+
```
Comment (DanielGuramulta) [inline src/hardware_trigger.cpp]: There are some extra lines here, also there is no empty line after the `}` of this method.

File: src/hardware_trigger.cpp [hunk 2]
```diff
@@ -183,6 +203,49 @@ HardwareTrigger::condition HardwareTrigger::digitalCondition(uint chnIdx)
...
+}
+bool HardwareTrigger::hasExternalTriggerIn()
+{
+	return true;
+}
+
+bool HardwareTrigger::hasCrossInstrumentTrigger()
+{
+	auto ret = iio_channel_find_attr(m_logic_channels[1], "logic_mode");
+	return true;
```
Comment (DanielGuramulta) [inline src/hardware_trigger.cpp]: What's the purpose of `ret` here, should it return `true`, or `(ret == NULL) ? false : true` like in `hasExternalTriggerOut()`?

File: src/hardware_trigger.cpp [hunk 3]
```diff
@@ -296,6 +359,17 @@ void HardwareTrigger::setSource(const QString& source)
  * Convenience function to be used when willing to use the trigger for only one
  * channel at a time.
  */
+
+bool HardwareTrigger::triggerIn() const
+{
+	return m_trigger_in;
+
+}
```
Comment (DanielGuramulta) [inline src/hardware_trigger.cpp]: No empty line after `}`

File: src/hardware_trigger.hpp [hunk 4]
```diff
@@ -36,6 +36,31 @@ class HardwareTrigger
 		N_DIGITAL_AND_ANALOG = 7,
 		N_DIGITAL_XOR_ANALOG = 8,
 	};
+	/*enum mode {
```
Comment (DanielGuramulta) [inline src/hardware_trigger.hpp]: Will `enum mode` still be needed in the future?

File: src/hardware_trigger.hpp [hunk 5]
```diff
@@ -85,18 +113,28 @@ class HardwareTrigger
 	void setStreamingFlag(bool);
 	bool getStreamingFlag();
 
+	bool hasExternalTriggerIn();
```
Comment (DanielGuramulta) [inline src/hardware_trigger.hpp]: The `bool has*` methods could be `const`.

File: src/logic_analyzer.cpp [hunk 6]
```diff
@@ -311,20 +317,21 @@ LogicAnalyzer::LogicAnalyzer(struct iio_context *ctx,
...
+
+	trigger_settings_ui->extern_to->setVisible(false);
 
 	/* Cursor Settings */
 	connect(trigger_settings_ui->cmb_trigg_extern_cond_1,
 		SIGNAL(currentIndexChanged(int)),
 		this, SLOT(setExternalTrigger(int)));
-	connect(trigger_settings_ui->cmb_trigg_extern_cond_2,
+
+	for(auto val : externalTriggerSourceMap) {
```
Comment (DanielGuramulta) [inline src/logic_analyzer.cpp]: The items in the `externalTriggerSourceMap` could be accessed by constant reference : `const auto &val`

File: src/logic_analyzer.cpp [hunk 7]
```diff
@@ -1693,16 +1713,16 @@ void LogicAnalyzer::updateBufferPreviewer()
...
-	trigger_settings_ui->cmb_trigg_extern_cond_2->setEnabled(enabled);
+
 	if( !enabled ) {
 		trigger_settings_ui->cmb_trigg_extern_cond_1->setCurrentIndex(0);
-		trigger_settings_ui->cmb_trigg_extern_cond_2->setCurrentIndex(0);
 		setHWTrigger(16, trigger_mapping[0]);
-		setHWTrigger(17, trigger_mapping[0]);
 	}
-	else {
+	else {
```
Comment (DanielGuramulta) [inline src/logic_analyzer.cpp]: Empty tab/spaces after `{`

File: src/trigger_settings.cpp [hunk 8]
```diff
@@ -112,11 +124,32 @@ TriggerSettings::TriggerSettings(std::shared_ptr<GenericAdc> adc,
...
 	ui->cmb_analog_extern->setCurrentIndex(0);
 	on_cmb_analog_extern_currentIndexChanged(0);
+
+	if(trigger->hasExternalTriggerIn())
+		ui->cmb_extern_src->addItem("External Trigger In");
+	if(trigger->hasCrossInstrumentTrigger())
+		ui->cmb_extern_src->addItem("Logic Analyzer");
+
+	if(trigger->hasExternalTriggerOut()){
+		for(auto val : externalTriggerOutMapping)
```
Comment (DanielGuramulta) [inline src/trigger_settings.cpp]: The items in the `externalTriggerOutMapping` could be accessed by constant reference : `const auto &val`

File: src/trigger_settings.cpp [hunk 9]
```diff
@@ -205,6 +254,11 @@ void TriggerSettings::setTriggerHysteresis(double hyst)
 		trigg_configs[current_channel].hyst_val = hyst;
 	}
 }
+/*
+void TriggerSettings::on_cmb_extern_src_currentIndexChanged(int index)
```
Comment (DanielGuramulta) [inline src/trigger_settings.cpp]: Is this still needed?

File: src/trigger_settings.cpp [hunk 10]
```diff
@@ -292,6 +358,18 @@ HardwareTrigger::mode TriggerSettings::determineTriggerMode(bool intern_checked,
 	return mode;
 }
 
+
+
```
Comment (DanielGuramulta) [inline src/trigger_settings.cpp]: Extra empty lines

File: src/trigger_settings.hpp [hunk 11]
```diff
@@ -23,7 +23,10 @@
 #include "hardware_trigger.hpp"
 
 #include <QWidget>
+#include <QButtonGroup>
```
Comment (DanielGuramulta) [inline src/trigger_settings.hpp]: Is `#include <QButtonGroup>` needed here?

File: src/hardware_trigger.cpp (hunk 1 above)
Comment (adisuciu) [reply src/hardware_trigger.cpp]: fixed

File: src/hardware_trigger.cpp (hunk 2 above)
Comment (adisuciu) [reply src/hardware_trigger.cpp]: test code slipped through - fixed

File: src/hardware_trigger.cpp (hunk 3 above)
Comment (adisuciu) [reply src/hardware_trigger.cpp]: fixed

File: src/hardware_trigger.hpp (hunk 4 above)
Comment (adisuciu) [reply src/hardware_trigger.hpp]: fixed - enabled the newer enum

File: src/hardware_trigger.hpp (hunk 5 above)
Comment (adisuciu) [reply src/hardware_trigger.hpp]: fixed

File: src/logic_analyzer.cpp (hunk 6 above)
Comment (adisuciu) [reply src/logic_analyzer.cpp]: fixed

File: src/logic_analyzer.cpp (hunk 7 above)
Comment (adisuciu) [reply src/logic_analyzer.cpp]: fixed

File: src/trigger_settings.cpp (hunk 8 above)
Comment (adisuciu) [reply src/trigger_settings.cpp]: fixed

File: src/trigger_settings.cpp (hunk 9 above)
Comment (adisuciu) [reply src/trigger_settings.cpp]: fixed

File: src/trigger_settings.cpp (hunk 10 above)
Comment (adisuciu) [reply src/trigger_settings.cpp]: fixed

File: src/trigger_settings.hpp (hunk 11 above)
Comment (adisuciu) [reply src/trigger_settings.hpp]: fixed

**Comment by adisuciu** [comment]
Daisychain should be enabled by default as it provides compensation for external trigger inputs - which are always enabled. However the widgets were not correctly initialized when a trigger out feature is not present. Fixed with https://github.com/analogdevicesinc/scopy/pull/693/commits/6c09fe465d33bc20ac9bd8ffbe2536429d1bf413

**Pull request merged**